modules/join/src/test/java/org/apache/ignite/internal/join/ClusterInitializerTest.java

```diff
@@ -0,0 +1,186 @@
+
+    @Mock
+    private TopologyService topologyService;
+
+    private ClusterInitializer clusterInitializer;
+
+    private final InitMessagesFactory msgFactory = new InitMessagesFactory();
+
+    @BeforeEach
+    void setUp(@Mock ClusterService clusterService) {
+        when(clusterService.messagingService()).thenReturn(messagingService);
+        when(clusterService.topologyService()).thenReturn(topologyService);
+
+        clusterInitializer = new ClusterInitializer(clusterService);
+    }
+
+    /**
+     * Tests the happy-case scenario of cluster initialization.
+     */
+    @Test
+    void testNormalInit() {
+        ClusterNode metastorageNode = new ClusterNode("metastore", "metastore", new NetworkAddress("foo", 123));
+        ClusterNode cmgNode = new ClusterNode("cmg", "cmg", new NetworkAddress("bar", 456));
+
+        when(topologyService.getByConsistentId(metastorageNode.name())).thenReturn(metastorageNode);
+        when(topologyService.getByConsistentId(cmgNode.name())).thenReturn(cmgNode);
+        when(topologyService.allMembers()).thenReturn(List.of(metastorageNode, cmgNode));
+
+        when(messagingService.invoke(any(ClusterNode.class), any(MetastorageInitMessage.class), anyLong()))
+                .thenAnswer(invocation -> {
+                    MetastorageInitMessage message = invocation.getArgument(1);
+
+                    NetworkMessage response = msgFactory.leaderElectedMessage()
+                            .leaderName(message.metastorageNodes()[0])
+                            .build();
+
+                    return CompletableFuture.completedFuture(response);
+                });
+
+        when(messagingService.invoke(any(ClusterNode.class), any(CmgInitMessage.class), anyLong()))
+                .thenAnswer(invocation -> {
+                    CmgInitMessage message = invocation.getArgument(1);
+
+                    NetworkMessage response = msgFactory.leaderElectedMessage()
+                            .leaderName(message.cmgNodes()[0])
+                            .build();
+
+                    return CompletableFuture.completedFuture(response);
+                });
+
+        // check that leaders are different in case different node IDs are provided
+        CompletableFuture<Leaders> leaders = clusterInitializer.initCluster(List.of(metastorageNode.name()), List.of(cmgNode.name()));
+
+        verify(messagingService).invoke(eq(metastorageNode), any(MetastorageInitMessage.class), anyLong());
+        verify(messagingService).invoke(eq(cmgNode), any(MetastorageInitMessage.class), anyLong());
+        verify(messagingService).invoke(eq(metastorageNode), any(CmgInitMessage.class), anyLong());
+        verify(messagingService).invoke(eq(cmgNode), any(CmgInitMessage.class), anyLong());
+
+        var expectedLeaders = new Leaders(metastorageNode.name(), cmgNode.name());
+
+        assertThat(leaders, willBe(equalTo(expectedLeaders)));
+
+        clearInvocations(messagingService);
```

Comment:
This usually means that a test is actually testing more than one behavior. I suggest splitting the test in two: one for the case when both meta-storage and CMG nodes are specified, another for the case when only meta-storage nodes are given.